packages/completer/src/handler.ts

```diff
@@ -477,6 +486,85 @@ export namespace CompletionHandler {
     connector: IDataConnector<IReply, void, IRequest>;
   }
 
+  /**
+   * Wrapper object for ICompletionItem.
+   * Implementers of this interface should be responsible for
+   * deduping and sorting the items in the list.
+   */
+  export interface ICompletionItems {
+    /**
+     * Collection of completion items.
+     */
+    items: Array<ICompletionItem>;
+  }
+
+  /**
+   * Completion item object based off of LSP CompletionItem
+   */
+  export interface ICompletionItem extends PartialJSONObject {
```

Comment:
> In the model, I use a couple of methods from the JSONExt library that require this interface to be a PartialJSONObject

Hm, I don't believe you have to extend it though for it to be compliant. Or you shouldn't have to!

Reply:
Can we just not have it allow any key? And if we need to pass it to a JSONExt thing, just cast it to a `PartialJSONObject`?

Reply:
> Can we just not have it allow any key?
```typescript
export interface ICompletionItem  {
    [key: any]: PartialJSONValue | undefined
```
does not work, since the index signature no longer matches `PartialJSONObject`, which requires keys as strings.

> And if we need to pass it to a JSONExt thing, just cast it to a PartialJSONObject?
Since we're passing in an array to `JSONExt`, we'd need to cast `ICompletionItems` into a `ReadonlyJSONArray`. Unfortunately, this would require the individual items to adhere to `PartialJSONValue` which itself is a type alias that encapsulates `PartialJSONObject`.

At a high level though, is the concern that the interface is too restrictive if we extend from `PartialJSONObject`, or force the keys of `ICompletionItem` to be strings?

Given that we care to use `JSONExt` for its `deepCopy` and deep equality functionality, there may be other libraries we could use apart from Lumino that gives us more type flexibility, if that's an avenue to explore.

Reply:
> At a high level though, is the concern that the interface is too restrictive if we extend 

No the concern is that it's too open ended, the opposite. I think the interface should *only* specify the keys you need from it, and not have any generic keys in it.

What do you need from `JSONExt`?

You could just cast when you pass it in (I know this isn't ideal, but it seems preferably then having an overly permissive public interface).

```typescript
function takesJSON(j: PartialJSONObject): void {

}

interface IMyInterface {
  someKey: string;
}

const i: IMyInterface = {someKey: "hi there"}

takesJSON(i as unknown  as PartialJSONObject);
```

Reply:
To give some more reasoning, my assumption is that if an interface allows arbitrary key/values I assume that those will be *used* in some way by the function that takes that interface. But that isn't the case here. We will only by using the the keys you manually specified. So it is a clearer contract to me with the callers if we specify that by not extending from the PartialJSONObject or allowing arbitrary keys.

Reply:
Thanks for the clarification! Casting to `unknown` first resolved the issues.

Reply:
Thank you! Thanks for bearing with my comments.